src/graph_visualization.cpp

```diff
@@ -0,0 +1,463 @@
+void removeIncidentArcs(std::vector<TreeNode> &nodes, TreeNode::nodeId id);
+void greedyFAS(std::vector<TreeNode> &nodes, std::vector<std::pair<TreeNode::nodeId, TreeNode::nodeId>> &deletedEdges);
+
+void algorithmASAP(std::vector<TreeNode> &nodes);
+
+void addDummyNode(std::vector<TreeNode> &nodes, TreeNode up, TreeNode down, int order);
+void addDummyNodes(std::vector<TreeNode> &nodes, std::vector<std::pair<TreeNode::nodeId, TreeNode::nodeId>> &deletedEdges);
+
+int main() {
+  std::vector<std::vector<int>> input = {};
+  inputGraph(input);
+
+  std::vector<TreeNode> nodes = {};
+  transformationGraph(input, nodes);
+
+  printInfo(nodes);
+
+  return 0;
+}
+
+void inputGraph(std::vector<std::vector<int>> &input) {
+  for (int i = 0; 1; i++) {
+    int inputNum = -3;
+    std::vector<int> in = {};
+    input.push_back(in);
+    while (1) {
+      scanf("%i", &inputNum);
+      if (inputNum == -1 || inputNum == -2) {
+        break;
+      }
+      input[i].push_back(inputNum);
+    }
+    if (inputNum == -2) {
+      break;
+    }
+  }
+}
+
+void initIdAndSucc(std::vector<std::vector<int>> &input, std::vector<TreeNode> &nodes) {
+  for (int i = 0; i < input.size(); i++) {
+    TreeNode node = {};
+    nodes.push_back(node);
+    nodes[i].id = i;
+    for (int j = 0; j < input[i].size(); j++) {
+      nodes[i].succ.push_back(input[i][j]);
+    }
+  }
+}
+
+void predReinit(std::vector<TreeNode> &nodes) {
+  for (int i = 0; i < nodes.size(); i++) {
+    nodes[i].pred = {};
+  }
+
+  for (int i = 0; i < nodes.size(); i++) {
+    for (int j = 0; j < nodes[i].succ.size(); j++) {
+      for (int k = 0; k < nodes.size(); k++) {
+        if (nodes[k].id == nodes[i].succ[j]) {
+          nodes[k].pred.push_back(nodes[i].id);
+        }
+      }
+    }
+  }
+}
+
+void degReinit(std::vector<TreeNode> &nodes) {
+  for (int i = 0; i < nodes.size(); i++) {
+    nodes[i].degP = 0;
+    nodes[i].degM = 0;
+  }
+
+  for (int i = 0; i < nodes.size(); i++) {
+    for (int j = 0; j < nodes[i].succ.size(); j++) {
+      nodes[i].degP++;
+
+      for (int k = 0; k < nodes.size(); k++) {
+        if (nodes[k].id == nodes[i].succ[j]) {
+          nodes[k].degM++;
+        }
+      }
+    }
+  }
+}
+
+void algorithmDFS(int &exit, std::vector<TreeNode> &nodes, std::vector<bool> &usedId, TreeNode::nodeId id, TreeNode::nodeId prevId) {
+  if (exit == 0) {
+    usedId[id] = true;
+
+    int numb = -1;
+    for (int i = 0; i < nodes.size(); i++) {
+      if (nodes[i].id == id) {
+        numb = i;
+        break;
+      }
+    }
+
+    for (int i = 0; i < nodes[numb].succ.size(); i++) {
+      if (!usedId[nodes[numb].succ[i]]) {
+        algorithmDFS(exit, nodes, usedId, nodes[numb].succ[i], id);
+      } else {
+        if (id != prevId) {
+          printf("Graph has cycle. \n");
+          exit = 1;
+          break;
+        }
+      }
+    }
+  }
+}
+
+void cycleExistsDFS(std::vector<TreeNode> &nodes) {
```

Comment:
Make the function return so we can actually use it in programs.
```suggestion
bool cycleExistsDFS(std::vector<TreeNode> &nodes) {
```